File: app/src/main/java/com/security/passwordmanager/data/DataViewModel.java
package com.security.passwordmanager.data;

import android.app.Application;
import android.content.ClipData;
import android.content.ClipboardManager;
import android.content.Context;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.lifecycle.AndroidViewModel;

import com.security.passwordmanager.Cryptographer;
import com.security.passwordmanager.R;

import java.util.ArrayList;
import java.util.List;

public class DataViewModel extends AndroidViewModel {

    private static final String COPY_LABEL = "DataViewModel_copy";

    private final DataRepository dataRepository;
    private final Application mApplication;
    private final Cryptographer mCryptographer;

    public DataViewModel(@NonNull Application application) {
        super(application);
        this.mApplication = application;
        mCryptographer = new Cryptographer(application);

        DataDao dataDao = MainDatabase.getDatabase(application).websiteDao();
        dataRepository = new DataRepository(dataDao);
    }

    public void addData(Data data) {
        dataRepository.addData(data.encrypt(mCryptographer));
    }

    public void updateData(Data data) {
        dataRepository.updateData(data.encrypt(mCryptographer));
    }

    public List<Data> getDataList() {
        List<Data> dataList = dataRepository.getDataList();
        decryptList(dataList);
        return dataList;
    }

    public List<Data> getAccountList(String key, @Data.DataType int type) {
        List<Data> accountList = dataRepository.getAccountList(key, type);
        if (accountList == null)
            return new ArrayList<>();
        else {
            decryptList(accountList);
            return accountList;
        }
    }

    public List<Data> searchData(String query) {
        return dataRepository.searchData(query);
    }

    public void deleteData(Data data) {
        dataRepository.deleteData(data.encrypt(mCryptographer));
    }

    public void deleteData(String key, @Data.DataType int type) {
        dataRepository.deleteData(key, type);
    }

    public void copyData(Data data) {
        String dataString = data.toString(mApplication, true);
        copyText(dataString);
    }

    public void copyAccountList(List<Data> accountList) {
        StringBuilder builder = new StringBuilder(
                accountList.get(0).toString(mApplication, true));

        for (int i = 1, accountListSize = accountList.size(); i < accountListSize; i++) {
            Data d = accountList.get(i);
            builder.append("\n").append(d.toString(mApplication, false));
        }

        copyText(builder.toString());
    }

    public void copyText(String text) {
        ClipboardManager clipboard = (ClipboardManager)
                mApplication.getSystemService(Context.CLIPBOARD_SERVICE);
        ClipData clip = ClipData.newPlainText(COPY_LABEL, text);
        clipboard.setPrimaryClip(clip);
        Toast.makeText(mApplication, R.string.clipText, Toast.LENGTH_SHORT).show();
    }




    private void decryptList(@NonNull List<Data> list) {
        for (int i = 0; i < list.size(); i++) {
            Data data = list.get(i);
            list.set(i, data.decrypt(mCryptographer));
        }
    }
}

File: app/src/main/java/com/security/passwordmanager/data/DataViewModel.kt
package com.security.passwordmanager.data

import android.app.Application
import android.content.ClipData
import android.content.ClipboardManager
import android.content.Context
import androidx.lifecycle.AndroidViewModel
import com.security.passwordmanager.Cryptographer
import com.security.passwordmanager.R
import com.security.passwordmanager.getString
import com.security.passwordmanager.settings.EnumPreferences.APP_PREFERENCES
import com.security.passwordmanager.settings.EnumPreferences.APP_PREFERENCES_LOGIN
import com.security.passwordmanager.showToast

class DataViewModel(application: Application) : AndroidViewModel(application) {

    companion object {
        private const val COPY_LABEL = "data copy"
    }

    private val mApplication = application
    private val dataRepository: DataRepository
    private val cryptographer = Cryptographer(application)

    private val email: String get() {
        val preferences = mApplication
            .getSharedPreferences(APP_PREFERENCES.title, Context.MODE_PRIVATE)

        return preferences.getString(APP_PREFERENCES_LOGIN.title) ?: ""
    }

    init {
        val dataDao = MainDatabase.getDatabase(application).dataDao()
        dataRepository = DataRepository(dataDao)
    }

    fun addData(data: Data) {
        data.email = email
        dataRepository.addData(data.encrypt(cryptographer::encrypt))
    }

    fun updateData(data: Data) =
        dataRepository.updateData(data.encrypt(cryptographer::encrypt))

    fun getDataList() =
        dataRepository.getDataList(email).decrypt()

    fun getAccountList(key: String, type: DataType) =
        dataRepository.getAccountList(email, key, type).decrypt()

    fun getAccountList(data: Data) = getAccountList(data.key, data.type)


    fun searchData(query: String, type: DataType? = null): List<Data> =
        dataRepository.searchData(email, query, type)

    fun deleteData(data: Data) = dataRepository.deleteData(data)

    fun deleteRecords(data: Data) = dataRepository.deleteRecords(data)

    fun copyText(text: String) {
        val clipboard = mApplication.getSystemService(Context.CLIPBOARD_SERVICE) as ClipboardManager
        val clip = ClipData.newPlainText(COPY_LABEL, text)
        clipboard.setPrimaryClip(clip)
        showToast(mApplication, R.string.clipText)
    }

    fun copyData(data: Data) {
        val dataString = data.toString(mApplication)
        copyText(dataString)
    }

    fun copyAccountList(accountList: List<Data>) {
        val builder = StringBuilder(
                accountList[0].toString(mApplication))

        for (index in 1 until accountList.size) {
            val d = accountList[index]
            builder.append("\n").append(d.toString(mApplication, false))
        }
        copyText(builder.toString())
    }



    private fun MutableList<Data>.decrypt(): List<Data> {
        for (i in indices) {
            val data = this[i]
            this[i] = data.decrypt(cryptographer::decrypt)
        }
        return this
    }
}